### ISSUE TITLE
'IPython.utils.io' has no attribute 'stdout' erro:

### ISSUE THREAD

**Issue by jian899** [ipython/ipython#14012]
Hi,

I'm using Python 3.8.16 with the last version of IPython (8.12.0) and I get the following error:

`AttributeError: module 'IPython.utils.io' has no attribute 'stdout'`

**Comment by Carreau** [comment]
Can you please post the full traceback ?

**Comment by jian899** [comment]
Ok,details，
```
~$ ipython --matplotlib
Python 3.8.16 (default, Jan 17 2023, 23:13:24)
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.0 -- An enhanced Interactive Python. Type '?' for help.
Installed qt5 event loop hook.
Shell is already running a gui event loop for qt5. Call with no arguments to disable the current loop.
Using matplotlib backend: Qt5Agg

In [1]: from gbm.data import Cspec
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/time/core.py:2594, in update_leap_seconds(files)
   2592 from astropy.utils import iers
-> 2594 table = iers.LeapSeconds.auto_open(files)
   2595 return erfa.leap_seconds.update(table)

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/utils/iers/iers.py:1025, in LeapSeconds.auto_open(cls, files)
   1024 if self.expires < self._today():
-> 1025     warn('leap-second file is expired.', IERSStaleWarning)
   1027 return self

File ~/miniconda3/envs/fermigbm/lib/python3.8/warnings.py:109, in _showwarnmsg(msg)
    106     raise TypeError("warnings.showwarning() must be set to a "
    107                     "function or method")
--> 109 sw(msg.message, msg.category, msg.filename, msg.lineno,
    110    msg.file, msg.line)
    111 return

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/logger.py:204, in AstropyLogger._showwarning(self, *args, **kwargs)
    203 if mod_name is not None:
--> 204     self.warning(message, extra={'origin': mod_name})
    205 else:

File ~/miniconda3/envs/fermigbm/lib/python3.8/logging/__init__.py:1458, in Logger.warning(self, msg, *args, **kwargs)
   1457 if self.isEnabledFor(WARNING):
-> 1458     self._log(WARNING, msg, args, **kwargs)

File ~/miniconda3/envs/fermigbm/lib/python3.8/logging/__init__.py:1589, in Logger._log(self, level, msg, args, exc_info, extra, stack_info, stacklevel)
   1587 record = self.makeRecord(self.name, level, fn, lno, msg, args,
   1588                          exc_info, func, extra, sinfo)
-> 1589 self.handle(record)

File ~/miniconda3/envs/fermigbm/lib/python3.8/logging/__init__.py:1599, in Logger.handle(self, record)
   1598 if (not self.disabled) and self.filter(record):
-> 1599     self.callHandlers(record)

File ~/miniconda3/envs/fermigbm/lib/python3.8/logging/__init__.py:1661, in Logger.callHandlers(self, record)
   1660     if record.levelno >= hdlr.level:
-> 1661         hdlr.handle(record)
   1662 if not c.propagate:

File ~/miniconda3/envs/fermigbm/lib/python3.8/logging/__init__.py:954, in Handler.handle(self, record)
    953 try:
--> 954     self.emit(record)
    955 finally:

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/logger.py:548, in StreamHandler.emit(self, record)
    547 elif record.levelno < logging.ERROR:
--> 548     color_print(record.levelname, 'brown', end='', file=stream)
    549 else:

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/utils/console.py:356, in color_print(end, *args, **kwargs)
    326 """
    327 Prints colors and styles to the terminal uses ANSI escape
    328 sequences.
   (...)
    353     be printed after resetting any color or font state.
    354 """
--> 356 file = kwargs.get('file', _get_stdout())
    358 write = file.write

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/utils/console.py:123, in _get_stdout(stderr)
    122 # Our system stream is an atty and we're in ipython.
--> 123 ipyio_stream = _IPython.get_stream(stream)
    125 if ipyio_stream is not None and isatty(ipyio_stream):
    126     # Use the IPython console output stream

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/utils/console.py:100, in _IPython.get_stream(cls, stream)
     98 @classmethod
     99 def get_stream(cls, stream):
--> 100     return getattr(cls.ipyio, stream)

AttributeError: module 'IPython.utils.io' has no attribute 'stdout'

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
Cell In[1], line 1
----> 1 from gbm.data import Cspec

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/gbm/data/__init__.py:3
      1 from .collection import DataCollection, GbmDetectorCollection
      2 from .drm import RSP
----> 3 from .localization import HealPix, GbmHealPix
      4 from .phaii import PHAII, Ctime, Cspec, TTE
      5 from .pha import PHA, BAK

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/gbm/data/localization.py:34
     31 from warnings import warn
     32 import warnings
---> 34 import healpy as hp
     35 from collections import OrderedDict
     36 from matplotlib.pyplot import contour as Contour

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/healpy/__init__.py:95
     92 from ._pixelfunc import ringinfo, pix2ring
     94 from ._sphtools import rotate_alm
---> 95 from .rotator import Rotator, vec2dir, dir2vec
     96 from ._healpy_pixel_lib import UNSEEN
     97 from .visufunc import (
     98     mollview,
     99     graticule,
   (...)
    107     azeqview,
    108 )

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/healpy/rotator.py:25
     21 import logging
     23 log = logging.getLogger("healpy")
---> 25 from astropy.coordinates import SkyCoord
     26 from astropy.utils.decorators import deprecated_renamed_argument
     27 from . import pixelfunc

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/coordinates/__init__.py:16
     14 from .earth import *
     15 from .transformations import *
---> 16 from .builtin_frames import *
     17 from .name_resolve import *
     18 from .matching import *

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/coordinates/builtin_frames/__init__.py:29
     27 from .baseradec import BaseRADecFrame
     28 from .icrs import ICRS
---> 29 from .fk5 import FK5
     30 from .fk4 import FK4, FK4NoETerms
     31 from .galactic import Galactic

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/coordinates/builtin_frames/fk5.py:8
      6 from astropy.coordinates.attributes import TimeAttribute
      7 from astropy.coordinates.transformations import DynamicMatrixTransform
----> 8 from astropy.coordinates import earth_orientation as earth
     10 from .baseradec import BaseRADecFrame, doc_components
     11 from .utils import EQUINOX_J2000

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/coordinates/earth_orientation.py:19
     15 from astropy import units as u
     16 from .matrix_utilities import rotation_matrix, matrix_product, matrix_transpose
---> 19 jd1950 = Time('B1950').jd
     20 jd2000 = Time('J2000').jd
     21 _asecperrad = u.radian.to(u.arcsec)

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/time/core.py:395, in Time.__new__(cls, val, val2, format, scale, precision, in_subfmt, out_subfmt, location, copy)
    389 if not _LEAP_SECONDS_CHECKED:
    390     # *Must* set to True first as update_leap_seconds uses Time.
    391     # In principle, this may cause wrong leap seconds in
    392     # update_leap_seconds itself, but since expiration is in
    393     # units of days, that is fine.
    394     _LEAP_SECONDS_CHECKED = True
--> 395     update_leap_seconds()
    397 if isinstance(val, cls):
    398     self = val.replicate(format=format, copy=copy)

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/time/core.py:2598, in update_leap_seconds(files)
   2595     return erfa.leap_seconds.update(table)
   2597 except Exception as exc:
-> 2598     warn("leap-second auto-update failed due to the following "
   2599          f"exception: {exc!r}", AstropyWarning)
   2600     return 0

File ~/miniconda3/envs/fermigbm/lib/python3.8/warnings.py:109, in _showwarnmsg(msg)
    105         if not callable(sw):
    106             raise TypeError("warnings.showwarning() must be set to a "
    107                             "function or method")
--> 109         sw(msg.message, msg.category, msg.filename, msg.lineno,
    110            msg.file, msg.line)
    111         return
    112 _showwarnmsg_impl(msg)

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/logger.py:204, in AstropyLogger._showwarning(self, *args, **kwargs)
    201         break
    203 if mod_name is not None:
--> 204     self.warning(message, extra={'origin': mod_name})
    205 else:
    206     self.warning(message)

File ~/miniconda3/envs/fermigbm/lib/python3.8/logging/__init__.py:1458, in Logger.warning(self, msg, *args, **kwargs)
   1449 """
   1450 Log 'msg % args' with severity 'WARNING'.
   1451
   (...)
   1455 logger.warning("Houston, we have a %s", "bit of a problem", exc_info=1)
   1456 """
   1457 if self.isEnabledFor(WARNING):
-> 1458     self._log(WARNING, msg, args, **kwargs)

File ~/miniconda3/envs/fermigbm/lib/python3.8/logging/__init__.py:1589, in Logger._log(self, level, msg, args, exc_info, extra, stack_info, stacklevel)
   1586         exc_info = sys.exc_info()
   1587 record = self.makeRecord(self.name, level, fn, lno, msg, args,
   1588                          exc_info, func, extra, sinfo)
-> 1589 self.handle(record)

File ~/miniconda3/envs/fermigbm/lib/python3.8/logging/__init__.py:1599, in Logger.handle(self, record)
   1592 """
   1593 Call the handlers for the specified record.
   1594
   1595 This method is used for unpickled records received from a socket, as
   1596 well as those created locally. Logger-level filtering is applied.
   1597 """
   1598 if (not self.disabled) and self.filter(record):
-> 1599     self.callHandlers(record)

File ~/miniconda3/envs/fermigbm/lib/python3.8/logging/__init__.py:1661, in Logger.callHandlers(self, record)
   1659     found = found + 1
   1660     if record.levelno >= hdlr.level:
-> 1661         hdlr.handle(record)
   1662 if not c.propagate:
   1663     c = None    #break out

File ~/miniconda3/envs/fermigbm/lib/python3.8/logging/__init__.py:954, in Handler.handle(self, record)
    952 self.acquire()
    953 try:
--> 954     self.emit(record)
    955 finally:
    956     self.release()

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/logger.py:548, in StreamHandler.emit(self, record)
    546     color_print(record.levelname, 'green', end='', file=stream)
    547 elif record.levelno < logging.ERROR:
--> 548     color_print(record.levelname, 'brown', end='', file=stream)
    549 else:
    550     color_print(record.levelname, 'red', end='', file=stream)

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/utils/console.py:356, in color_print(end, *args, **kwargs)
    325 def color_print(*args, end='\n', **kwargs):
    326     """
    327     Prints colors and styles to the terminal uses ANSI escape
    328     sequences.
   (...)
    353         be printed after resetting any color or font state.
    354     """
--> 356     file = kwargs.get('file', _get_stdout())
    358     write = file.write
    359     if isatty(file) and conf.use_color:

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/utils/console.py:123, in _get_stdout(stderr)
    120     return sys_stream
    122 # Our system stream is an atty and we're in ipython.
--> 123 ipyio_stream = _IPython.get_stream(stream)
    125 if ipyio_stream is not None and isatty(ipyio_stream):
    126     # Use the IPython console output stream
    127     return ipyio_stream

File ~/miniconda3/envs/fermigbm/lib/python3.8/site-packages/astropy/utils/console.py:100, in _IPython.get_stream(cls, stream)
     98 @classmethod
     99 def get_stream(cls, stream):
--> 100     return getattr(cls.ipyio, stream)

AttributeError: module 'IPython.utils.io' has no attribute 'stdout'
```

**Comment by Carreau** [comment]
IPython.utils.io.stdout has been remove in december 2021 since IPython 8.0 and  deprecated since IPython 6.0.

Its usage has been remove in astropy since https://github.com/astropy/astropy/pull/10942  (octobre 2020) released in version 4.3 (release july 2021).


You want to upgrade astropy or downgrade IPython.
